### PR TITLE
[performance] block device test

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Reason for This PR

Lack of a block device performance test.

## Description of Changes

Added a test for Firecracker block device performance regression detection:
* Used `fio` to compute: iops, bandwidth, latency.
* Used `top` to compute cpu utilization host/guest.
* `fio` workloads cover the following modes: randread, randrw, read, readwrite.
* Block sizes for disk requests vary between: 512, 1024, 4096, 65536 bytes.

**NOTE**: I am still working through validating the computed baselines.

~~- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).~~

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
~~- [ ] Any newly added `unsafe` code is properly documented.~~
~~- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.~~
~~- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.~~
~~- [ ] All added/changed functionality is tested.~~

### CODE REVIEW

**Comment by alindima** [comment]
we should also modify or check the host block device scheduler. Having a different scheduler would also influence the results

File: tests/framework/statistics/consumer.py
Comment (raduiliescu) [inline tests/framework/statistics/consumer.py:80]: For the current PR this is ok, but I am not sure if this is the best way in terms of extensibility. This function will either return true or assert. There is no case for false. Maybe in the future you will want to reuse it, and this behaviors is quite strict. I would use a print or return a string, where empty string will be ok. This way the upper level function can decide if it wants to assert or not.

I am not blocking the PR on this. Is just food for thought.

File: tests/integration_tests/performance/test_block_performance.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_block_performance.py:1]: Copyright 2020

File: tests/integration_tests/performance/test_block_performance.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_block_performance.py:154]: Can't this return list statements made generic, like a matrix where you say how do you want to calculate each param and then you just walk through the matrix and add the stuff? The code here is pretty hard to read.

File: tests/integration_tests/performance/test_block_performance.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_block_performance.py:174]: Don't fully understand the differences between if cpu and if not cpu.  Can you please explain?

File: tests/integration_tests/performance/test_block_performance.py
Comment (alindima) [reply tests/integration_tests/performance/test_block_performance.py:340]: it's a dict, so `thread_id = list(cpu_load[tag])[0]`

File: tests/framework/statistics/consumer.py
Comment (alindima) [reply tests/framework/statistics/consumer.py:80]: Another way would be to have an extra configuration to this function -> `assert=True`

File: tests/integration_tests/performance/test_block_performance.py
Comment (alindima) [reply tests/integration_tests/performance/test_block_performance.py:23]: Tried it. Unfortunately, using `import *`, pylint will give us a warning. This could be avoided by adding this comment: `# pylint: disable=W0401`.

This is not enough, since pytlint also throws errors, stating that every indentifier in the imported file is undefined. We could mitigate by this adding another supression comment.

However, I wouldn't do it, since this would make pylint ignore any other undefined variable error (some that should actually be errors, e.g.: `print(sadasnjkldsfndsksncksd)`

File: tests/integration_tests/performance/block_performance_test_config.py
Comment (gc-plp) [reply tests/integration_tests/performance/block_performance_test_config.py:1]: We should not store them at all in this format. 

It's difficult to read and difficult to update the values.

File: tests/integration_tests/performance/test_block_performance.py
Comment (gc-plp) [reply tests/integration_tests/performance/test_block_performance.py:23]: Just use `from integration_tests.performance.block_performance_test_config as test_config`.

Then you can use `test_config.FIO.something`.

File: tests/integration_tests/performance/test_block_performance.py
Comment (gc-plp) [inline tests/integration_tests/performance/test_block_performance.py:74]: Shouldn't we have a generic function for measuring CPU utilization and feeding the expected values?

I don't really understand why we need to define CPU usage comparisons criteria since these should be universal.

File: tests/integration_tests/performance/test_block_performance.py
Comment (gc-plp) [inline tests/integration_tests/performance/test_block_performance.py:77]: I don't think that using nested dicts is the way to go here.

We should rethink this.

File: tests/integration_tests/performance/test_block_performance.py
Comment (gc-plp) [inline tests/integration_tests/performance/test_block_performance.py:180]: This block of code is hard to read and understand due to the mix of indentations and multiple elements being extended.

File: tests/integration_tests/performance/test_block_performance.py
Comment (gc-plp) [reply tests/integration_tests/performance/test_block_performance.py:23]: Also, as a rule of thumb, keep imported symbols to a minimum. If multiple symbols need to be imported from one file, use the import name as a prefix as stated above.

Otherwise we add too many symbols in the current module and their origins become hard to track.

**Review by gc-plp — COMMENTED** [review]
The overall test achieves its goal, but there are a few things to consider before merging this. I left a few comments above, but I think that a summary is needed:

1. The huge block_performance_test_config.py file needs to be changed into something else. Ideally this should act as a backing store for a class based approach that reads this data. Storing it as a Python file that also needs nice formatting to make the linters play cool is not creating an easy to maintain code base.
We should explore other options here. A CSV file perhaps? Breaking it down into multiple files?

2. Repeated use of nested dicts. Use carefully. Ideally use a class based approach.

3. A file that contains a test or tests should only define the high level steps that the test does in a very clear way. Other dependencies should be layered either in the framework that supports the test or secondary test files that do not define tests.

4. The statistics framework itself needs a lot of explicit elements that it has to be given. For example having to always append things to the `StatisticDef` member. Could the framework identify what elements it needs by reading the test baseline definition?
I'm thinking that the huge statistics file could instead define a bunch of key variables that tell the framework how the tata in that file is read and used. Right now we always need to trickle down information from the top level, but that information is very specific to test cases and could be also built from the ground up.

File: tests/integration_tests/performance/block_performance_test_config.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/block_performance_test_config.py]: Yes. I am trying to merge baselines in baby steps. Also, 4k block size is the most common in disks benchmarking. I think we can assume this is the goto relevant measurement to be used to assess virtio block device performance, for regression analysis.

File: tests/integration_tests/performance/block_performance_test_config.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/block_performance_test_config.py]: I am not aware of any reasons whatsoever. This test was adapted from a previous existing one, which was not interested in write only benchmarking. Having dedicated cases for other scenarios might be a follow up after merging this test, which is the foundation really. The scenarios can be really diverse, so it is a matter of what we find interesting or relevant for customers use cases.

File: tests/host_tools/network.py
Comment (alindima) [inline tests/host_tools/network.py:63]: should it be `"""Copy a file from the VM using scp"""`? I see that there is already a function in this file that handles copying to the VM

File: tests/integration_tests/performance/test_vsock_throughput.py
Comment (alindima) [inline tests/integration_tests/performance/test_vsock_throughput.py:39]: maybe we should rename the function. the stats are not basic, they just don't have pass criteria

File: tests/integration_tests/performance/test_vsock_throughput.py
Comment (alindima) [inline tests/integration_tests/performance/test_vsock_throughput.py:55]: i'd say to remove this function and just paste the code in the caller, the same way as before.
either that or rename it to `stats_with_criteria` (and the `basic_vsock_stats` function to `stats_without_criteria`)

**Review by alindima — COMMENTED** [review]
Great work with the framework refactoring 🚀

File: tests/framework/statistics/core.py
Comment (raduiliescu) [reply tests/framework/statistics/core.py:72]: That would be really nice, but if that is not trivial to do, can we make it in a followup PR.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (gc-plp) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:21]: Can we just use
```python
import integration_tests.performance.configs.network_tcp_throughput_test_config as test_cfg
```

And further down just use `test_cfg.THROUGHPUT` and so on.

WDYT?

File: tests/framework/statistics/core.py
Comment (iulianbarbu) [reply tests/framework/statistics/core.py:72]: Yes. I would focus on this in another PR.

File: tests/conftest.py
Comment (gc-plp) [inline tests/conftest.py:91]: Reduce clutter by importing `framework.defs` as an alias.

File: tests/framework/defs.py
Comment (gc-plp) [inline tests/framework/defs.py:34]: This variable definition is not robust as it references a different path than before, introducing breaking changes in the CI suite (e.g. gitlint test).

File: tests/integration_tests/performance/configs/block_performance_test_config_raw.json
Comment (alindima) [inline tests/integration_tests/performance/configs/block_performance_test_config_raw.json:789]: add a newline here

File: tests/integration_tests/performance/test_block_performance.py
Comment (alindima) [reply tests/integration_tests/performance/test_block_performance.py]: I see, the names are clearer now 👍🏻

**Review by iulianbarbu — COMMENTED** [review]
@gc-plp  and @alindima , I've managed to produce stable baselines and to address your comments. Please take a look again! Thanks for the patience. :)

File: tests/integration_tests/performance/test_block_performance.py
Comment (alindima) [inline tests/integration_tests/performance/test_block_performance.py:48]: this constant does not seem to be used

**Review by alindima — COMMENTED** [review]
Looks good, just some minor comments

File: tests/conftest.py
Comment (iulianbarbu) [reply tests/conftest.py:97]: Yes. PTAL!

File: tests/conftest.py
Comment (iulianbarbu) [reply tests/conftest.py:180]: Done.

File: tests/integration_tests/performance/test_block_performance.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_block_performance.py:48]: Done.

**Review by alindima — APPROVED** [review]
LGTM, great job 🚀

**Review by gc-plp — APPROVED** [review]

**Pull request merged**
